### PR TITLE
Add option to disable auto-update

### LINKED ISSUE #62: Add a configuration setting to disable auto-update
Since some corporate users can't auto-update their agent, auto-updating the client can cause breakages for them. 
Add a separate config value where they can specify the exact version they want their IDE to download - an empty value meaning just use whatever's there at the moment.
If no CLI has been downloaded and auto-update is disabled, download the latest version.

### PR DESCRIPTION
**Pull request by infiniteregrets**
closes #62

### CODE REVIEW

File: src/status.ts [hunk 1]
```diff
@@ -37,7 +39,8 @@ export class MirrordStatus {
         const activeConfig = MirrordConfigManager.getInstance().activeConfig();
         if (activeConfig) {
             statusBar.tooltip.appendMarkdown(`\n\n[Active config: ${vscode.workspace.asRelativePath(activeConfig)}](${activeConfig})`);
-        }
+        }               
+        statusBar.tooltip.appendMarkdown(`\n\n[Auto-update: ${autoUpdate ? 'Enabled' : 'Disabled'}](command:${MirrordStatus.autoUpdateCommandId})`);
```
Comment (eyalb181) [inline src/status.ts:43]: We don't want this in the menu, or as a command - it should only be available in the Extension Settings.

File: src/binaryManager.ts [hunk 2]
```diff
@@ -98,27 +86,44 @@ async function getConfiguredMirrordBinary(): Promise<string | null> {
...
-    if (localMirrord) {
-        const api = new MirrordAPI(localMirrord);
-        const installedVersion = await api.getBinaryVersion();
-        if (installedVersion === latestVersion) {
-            return localMirrord;
+    const extensionMirrordPath = getExtensionMirrordPath();
+    const latestVersion = await getLatestSupportedVersion(10000);
+
+    if (autoUpdate) {
+        await downloadMirrordBinary(extensionMirrordPath, latestVersion);
```
Comment (eyalb181) [inline src/binaryManager.ts:97]: What about the check that was done before, i.e. that maybe the binary is already installed in the correct version?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: src/status.ts (hunk 1 above)
Comment (infiniteregrets) [reply src/status.ts:43]: oh okay, do you mean something like

```
mirrord.autoUpdate": true
```
in workspace settings.json?

Comment (eyalb181) [reply src/status.ts:43]: Yeah, see how binaryPath is implemented.

File: test-workspace/.vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+    "mirrord.autoUpdate": {
+        "enabled": false,
+        "version": "3.70.0"
```
Comment (eyalb181) [inline test-workspace/.vscode/settings.json:4]: Remove?

File: src/extension.ts [hunk 4]
```diff
@@ -11,6 +12,9 @@ export async function activate(context: vscode.ExtensionContext) {
 
 	context.workspaceState.update('enabled', false);
 	vscode.debug.registerDebugConfigurationProvider('*', new ConfigurationProvider(), 2);
+	
+	// start mirrord binary updates, so that we avoid downloading mid session
+	await getMirrordBinary(true);
```
Comment (eyalb181) [inline src/extension.ts:17]: @Razz4780 Was there a reason we didn't do this originally?

File: src/binaryManager.ts [hunk 5]
```diff
@@ -19,37 +21,42 @@ function getExtensionMirrordPath(): Uri {
...
 /**
  * Tries to find local mirrord in path or in extension storage.
+ * @param version If specified, then the version of the binary is checked and matched path is returned if it matches.
+ * @returns Path to mirrord binary or null if not found
  */
 export async function getLocalMirrordBinary(version?: string): Promise<string | null> {
     try {
-        const mirrordPath = await which("mirrord");
+        const mirrordPath = getExtensionMirrordPath();
+        await workspace.fs.stat(mirrordPath);
```
Comment (eyalb181) [inline src/binaryManager.ts:30]: Why is this necessary?

File: src/binaryManager.ts [hunk 6]
```diff
@@ -19,37 +21,42 @@ function getExtensionMirrordPath(): Uri {
 
 /**
  * Tries to find local mirrord in path or in extension storage.
+ * @param version If specified, then the version of the binary is checked and matched path is returned if it matches.
+ * @returns Path to mirrord binary or null if not found
  */
 export async function getLocalMirrordBinary(version?: string): Promise<string | null> {
     try {
-        const mirrordPath = await which("mirrord");
+        const mirrordPath = getExtensionMirrordPath();
```
Comment (eyalb181) [inline src/binaryManager.ts:29]: Why'd you switch around the order of the code in this function, i.e. checking in extension storage first?

File: src/binaryManager.ts [hunk 7]
```diff
@@ -93,33 +100,76 @@ async function getConfiguredMirrordBinary(): Promise<string | null> {
...
-    const latestVersion = await getLatestSupportedVersion(timeout);
-
-    // See if maybe we have it installed already, in correct version.
-    const localMirrord = await getLocalMirrordBinary(latestVersion);
-    if (localMirrord) {
-        const api = new MirrordAPI(localMirrord);
-        const installedVersion = await api.getBinaryVersion();
-        if (installedVersion === latestVersion) {
-            return localMirrord;
+    // if extension is activating, then set a global state to what was read in the workspace settings
```
Comment (eyalb181) [inline src/binaryManager.ts:126]: Why?

File: src/binaryManager.ts [hunk 8]
```diff
@@ -93,33 +100,76 @@ async function getConfiguredMirrordBinary(): Promise<string | null> {
...
+                    await vscode.commands.executeCommand("workbench.action.reloadWindow");
+                })
+                .warning();
         }
+        return extensionPath;
     }
 
-    const extensionMirrordPath = getExtensionMirrordPath();
-    await downloadMirrordBinary(extensionMirrordPath, latestVersion);
+    if (typeof autoUpdateConfigured === 'string') {
```
Comment (eyalb181) [inline src/binaryManager.ts:143]: Document this please, i.e. possible values for autoUpdate

File: src/binaryManager.ts [hunk 9]
```diff
@@ -93,33 +100,76 @@ async function getConfiguredMirrordBinary(): Promise<string | null> {
...
+ * @returns Path to mirrord binary
+*/
+export async function getMirrordBinary(extensionActivate?: boolean): Promise<string> {
     const configured = await getConfiguredMirrordBinary();
+
     if (configured) {
+        vscode.window.showInformationMessage(`Using mirrord binary specified in settings: ${configured}`);
         return configured;
     }
+    const latestVersion = await getLatestSupportedVersion(10000);
```
Comment (eyalb181) [inline src/binaryManager.ts:122]: Before, the code had logic where if there was a local binary already, the timeout was shorter. Why was it removed?

File: src/binaryManager.ts [hunk 10]
```diff
@@ -93,33 +100,76 @@ async function getConfiguredMirrordBinary(): Promise<string | null> {
...
+        if (!autoUpdateConfigured) {
+            const localMirrordBinary = await getLocalMirrordBinary();
+            if (localMirrordBinary) {
+                return localMirrordBinary;
+            }
+            return extensionPath;
+        }
+    }
+
+    // Auto-update is enabled or no specific version specified.
```
Comment (eyalb181) [inline src/binaryManager.ts:166]: wdym "no specific version specified"? In what condition will this code be reached if auto update isn't enabled?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: src/binaryManager.ts (hunk 6 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:29]: check extension storage first so that if we downloaded a binary we use that first and not in path

File: src/binaryManager.ts (hunk 5 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:30]: i did not add this, existed before, i guess it just makes sure that it exists and the exception is caugght

File: src/binaryManager.ts (hunk 7 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:126]: mainly because we want to prompt the user to reload the IDE to avoid session timing out mid session

File: src/binaryManager.ts (hunk 6 above)
Comment (eyalb181) [reply src/binaryManager.ts:29]: Why wasn't this the order before? What changed?

File: src/binaryManager.ts (hunk 10 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:166]: old comment, please ignore

File: src/binaryManager.ts (hunk 6 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:29]: I am not sure why, i believe we should check extension path first

File: src/binaryManager.ts (hunk 9 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:122]: I dont get the reasoning behind the timeout, was gonna ask Michal tmrw

File: src/binaryManager.ts (hunk 8 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:143]: do you want me to add docstring for it? its already documented in package.json as json schema

File: src/binaryManager.ts (hunk 6 above)
Comment (eyalb181) [reply src/binaryManager.ts:29]: Why'd the person who wrote the original code believe otherwise? Ping them and make sure there's nothing we're overlooking

File: src/binaryManager.ts (hunk 8 above)
Comment (eyalb181) [reply src/binaryManager.ts:150]: It should be clear when reading the code without referring to package.json

File: src/binaryManager.ts (hunk 7 above)
Comment (eyalb181) [reply src/binaryManager.ts:133]: Why would a session time out when the user didn't reload the IDE?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: src/binaryManager.ts (hunk 7 above)
Comment (infiniteregrets) [reply src/binaryManager.ts:133]: user specifies new binary version -> starts debug session -> binary is downloading -> debug session times out

IMO instead now any changes different from original workspace settings pertaining to mirrord require reload

File: src/binaryManager.ts (hunk 6 above)
Comment (Razz4780) [reply src/binaryManager.ts:29]: I believe we were originally checking `PATH` before the extension storage for our own convenience (make changes in the CLI -> add to `PATH` -> works). Given that we always check the version of binaries we find and that now we have other means to use preferred binary (e.g. by setting binary path in extension settings), I don't think the order matters at all

File: src/binaryManager.ts (hunk 9 above)
Comment (Razz4780) [reply src/binaryManager.ts]: Originally that was Aviram's code, I didn't analyse. At some point I was just moving stuff around

Comment (infiniteregrets) [reply src/binaryManager.ts]: I have removed it since we were not using it

File: src/binaryManager.ts (hunk 7 above)
Comment (eyalb181) [reply src/binaryManager.ts:143]: The feature for specifying a version is *very* niche and we only expect users to use it when we instruct them to, so the edge case you're describing doesn't really matter. I'd rather keep the code simple.

Comment (infiniteregrets) [reply src/binaryManager.ts:143]: okay i a removing it

File: src/extension.ts (hunk 4 above)
Comment (eyalb181) [reply src/extension.ts]: @infiniteregrets So why is it necessary now?

Comment (infiniteregrets) [reply src/extension.ts]: because downloading mid-session timed out the debug session for me

Comment (infiniteregrets) [reply src/extension.ts]: I rested a few times with a manual timeout, its not happening anymore now? might be something else that I encountered

Comment (infiniteregrets) [reply src/extension.ts]: I think we can still keep this change, as it still helps make the extension faster in most cases, @Razz4780 wdyt

**Review by eyalb181 — APPROVED** [review]
